### ISSUE TITLE
Request Clarification on Evaluation

### ISSUE THREAD

**Issue by bhsimon0810** [Lydia-yang/FedDPA#1]
Hi, 

Congratulations on your excellent work! I have been following your work with great interest, and I would appreciate some clarifications regarding the evaluation process. 

To evaluate the fine-tuned model, I should first run the `GlobalModel_generate .py` script to get the predictions, and then use `metric.py` to evaluate the performance, am I correct?

Thank you in advance for your guidance!

Best regards,
Hao

**Comment by Lydia-yang** [comment]
Thank you for your interest. Yes, to evaluate the performance, first run GlobalModel_generate.py to obtain the predictions, and then run metric.py to compare the predictions with the ground-truth answers for the final results.

**Comment by bhsimon0810** [comment]
Thanks for your reply! 

Following the README, I use the exact scripts for both finetuning and inference:
```bash
# finetune
python main.py --global_model 'meta-llama/Llama-2-7b-hf'\
      --data_path  "./data/dataset1" \
      --output_dir  './lora-7b/'\
      --num_communication_rounds 20 \
      --num_clients  8 \
      --prompt_template_name 'alpaca_short' \
      --client_selection_frac 1 \
      --local_model True  # if true, learn global and local adapters iteratively (FedDPA-T)

# inference
python GlobalModel_generated.py \
      --load_8bit \
      --base_model 'meta-llama/Llama-2-7b-hf' \
      --lora_weights_path lora-7b/sft_model  \
      --lora_config_path lora-7b/sft_model   \
      --prompt_template 'alpaca_short' \
      --output_file 'output/result.jsonl' \
      --test_file './data/dataset1/flan_test_200_selected_nstrict_1.jsonl'
      
```

Then, to evaluate the performance, I ran the `metric.py` with the following modifications (Line 67-69):
```Python
targets = read_list('data/dataset1/flan_test_200_selected_nstrict_1.jsonl', 'output')
predictions = read_list('output/result.jsonl', 'answer')  # generated in the inference stage
get_result(targets, predictions, 'output/eval.json')
```
When running `metric.py`, I encountered some compatibility issues with Python 3.8, so I switched to Python 3.10 to execute this script. However, the evaluation results I obtained were significantly lower than those reported in your paper.

May I kindly ask if you could share the exact fine-tuning and inference scripts or configurations used to produce the reported results? Also, do you think switching to Python 3.10 for running `metric.py` might affect the evaluation results?

Thank you in advance for your time and clarification!

**Comment by Lydia-yang** [comment]
Sure, the scripts are provided below.
```
# finetune
python main.py --global_model 'meta-llama/Llama-2-7b-hf'\
      --data_path  "./data/dataset1" \
      --output_dir  './lora-7b/'\
      --num_communication_rounds 21 \
      --num_clients  8 \
      --prompt_template_name 'alpaca_short' \
      --client_selection_frac 1 \
      --local_model True  \ # if true, learn global and local adapters iteratively (FedDPA-T)
      --local_num_epochs 10 \
      --local_micro_batch_size 32 \ 
      --local_weight 0.5 # adjust with different dataset

# inference
for((i=0;i<=7;i++));
do
python GlobalModel_generated.py \
      --load_8bit \
      --base_model 'meta-llama/Llama-2-7b-hf' \
      --lora_weights_path "lora-7b/sft_model/20/local_output_${i}/local/pytorch_model.bin" \ # set local adapter
      --lora_config_path 'lora-7b/sft_model'   \
      --prompt_template 'alpaca_short' \
      --output_file 'output/result.jsonl' \
      --test_file './data/dataset1/flan_test_200_selected_nstrict_1.jsonl' \
      --batched False \
      --local True \
      --local_model_path "lora-7b/sft_model/19/adapter_model.bin"  \ # set global adapter
      --input_file "./data/dataset1/8/local_training_${i}.json"  \ 
      --auto True \ # instance-wise dynamic weighting mechanism
      --max_weight 0.5 # align with local_weight during finetune
done
```

I believe the main reason for the lower evaluation results is the inference configuration. If you only use the global adapter during inference, the results will be equivalent to FedIT. However, FedDPA-T requires both the global and local adapters for proper inference, which means you need to run the inference script for each client individually (i.e., 8 times for our dataset with 8 clients). The reported evaluation results are the test-time personalization performance for each client.

I don’t think the Python environment would have a significant impact on the results, but the random seed may affect. You may consider running the script multiple times to observe the performance variation and select the best result.

**Comment by bhsimon0810** [comment]
Thank you so much for your clarification! I am now able to fine-tune and evaluate the model successfully. 

By the way, I noticed that in `metric.py`, in addition to the ROUGE scores for each task, there is also a score named "total". Just to confirm, this score is not used for reporting and is different from the "Average" reported in Table 1&2 of the paper. Am I correct?

**Comment by Lydia-yang** [comment]
Yes, we do not report the "total" but calculate the average of all clients' scores to report "Average".

**Comment by bhsimon0810** [comment]
Thanks for your clarification!
